fplll/householder.cpp

```diff
@@ -0,0 +1,590 @@
+      V(i, k) = 0.0;
+
+// Setting R(i, k) to 0 is not neccessary since this value will be not used in HLLL.
+// However, in DEBUG, we must set to do not break test.
+#ifdef DEBUG
+      R(i, k) = 0.0;
+#endif  // DEBUG
+    }
+
+#ifdef HOUSEHOLDER_PRECOMPUTE_INVERSE
+    // Result is inf.
+    // TODO: set inf instead of doing the computation.
+    R_inverse_diag[i].div(1.0, 0.0);
+#endif  // HOUSEHOLDER_PRECOMPUTE_INVERSE
+  }
+
+  n_known_rows++;
+}
+
+// This function corresponds more or less to householder_r in hplll.
+// TODO: in hplll, householder_r do not recompute all the coefficients, depending on various
+// strategies. It will help to improve the running time.
+template <class ZT, class FT> void MatHouseholder<ZT, FT>::update_R(int i, bool last_j)
+{
+  // To update row i, we need to know n_known_rows rows
+  FPLLL_DEBUG_CHECK(i <= n_known_rows);
+  if (/*i == n_known_rows && =*/!updated_R)
+  {
+    int j, k;
+
+    for (j = 0; j < i; j++)
+    {
+      // vj * ri[j..n]^T
+      // TODO: n_known_cols here?
+      V[j].dot_product(ftmp0, R[i], j, n);
+
+      //-vj * ri[j..n]^T
+      ftmp0.neg(ftmp0);
+      // ri[j..n] = ri[j..n] - (vj * ri[j..n]^T) * vj
+      R[i].addmul(V[j], ftmp0, j, n);
+
+      // ri[j] = sigma[j] * ri[j]
+      R(i, j).mul(sigma[j], R(i, j));
+
+      // Copy R into R_history
+      // TODO: this is a row operation. Can try to copy R(i, j..n-1) directly in
+      // R_history[i][j](j..n-1).
+      for (k               = j; k < n; k++)
+        R_history[i][j][k] = R(i, k);
+    }
+
+    // Compute R(i, i), since R(i, 0..i-1) are known
+    if (last_j)
+      update_R_last(i);
+  }
+}
+
+template <class ZT, class FT> void MatHouseholder<ZT, FT>::refresh_R_bf(int i)
+{
+  // Verify that b[i] has change before last time (more or less bf != b)
+  FPLLL_DEBUG_CHECK(col_kept[i] == false);
+
+  int j;
+
+  n_known_cols = max(n_known_cols, init_row_size[i]);
+
+  if (enable_row_expo)
+  {
+    long max_expo = LONG_MIN;
+
+    // Copy b(i, j) in bf(i, j) and get the maximal exponent of the row
+    // TODO: these are row operations. However, since the types are converted, maybe difficult
+    // to make a row operation here.
+    for (j = 0; j < n_known_cols; j++)
+    {
+      b(i, j).get_f_exp(bf(i, j), tmp_col_expo[j]);
+      max_expo = max(max_expo, tmp_col_expo[j]);
+    }
+
+    // Renormalize all the bf(i, j) with max_expo
+    // TODO: these are row operations. However, since the exponent of mul_2si is not constant,
+    // maybe difficult to make a row operation here.
+    for (j = 0; j < n_known_cols; j++)
+      bf(i, j).mul_2si(bf(i, j), tmp_col_expo[j] - max_expo);
+    for (j = n_known_cols; j < n; j++)
+      bf(i, j) = 0.0;
+
+    row_expo[i] = max_expo;
+    FPLLL_DEBUG_CHECK(row_expo[i] >= 0);
+  }
+  else
+  {
+    // Simply copy b[i] in bf[i]
+    // TODO: these are row operations. However, since the types are converted, maybe difficult
```

Comment:
What is the status of this TODO?

Reply:
At one point, I decided to try to write row operations in `matrix.h` instead of explicitly write the for loop on the element in `householder.cpp`. However, some of these operations can be shared by other part of the code in fplll, in `mat_gso*` for example, and I do not take the time to factorize all the code at this step, but this is a project I plan to do in a near future. So for now, this is still a TODO.